File: backend/insights.py
"""Sorumlunun sistemi izleyip iyilestirmesi icin uc sinyal:
  1) yanit kalitesi     - basariyla cevaplanmis sorularin guven dagilimi
  2) insana yonlendirme - kanit yetersizligiyle sorumluya devredilen soru orani
  3) sik sorulan konular - HENUZ CEVAPLANAMAMIS, sorumluya yonlenen sorular
     arasinda anlamca (kelime kelime degil) tekrar eden kumeler

Kumeleme icin RAG'de zaten yuklu olan embedding modeli yeniden kullanilir -
ek bagimlilik veya ayri bir servis gerekmez.
"""

import time

from local_embed import embed_passages
from local_rag_answer import CONFIDENCE_HIGH

FREQUENT_MIN_COUNT = 2
FREQUENT_SIMILARITY = 0.86
FREQUENT_TOP_N = 20


def quality_breakdown(log):
    """Basariyla cevaplanmis sorularin guven seviyesine gore dagilimi."""
    answered = [e for e in log if e["status"] == "answered" and e.get("top_score") is not None]
    high = sum(1 for e in answered if e["top_score"] > CONFIDENCE_HIGH)
    return {"high": high, "mid": len(answered) - high, "total": len(answered)}


def recent_cutoff(days):
    """'simdi - days gun' zaman damgasi (log formatiyla: 'YYYY-MM-DD HH:MM:SS')."""
    return time.strftime("%Y-%m-%d %H:%M:%S", time.localtime(time.time() - days * 86400))


def referral_rate(log, since=None, until=None):
    """[since, until) araligindaki kayitlarda insana yonlendirme orani.
    Zaman damgalari sabit genislikte oldugundan sozlukce (lexical) karsilastirma
    kronolojik karsilastirmayla ayni sonucu verir; ayristirmaya gerek yoktur."""
    subset = [
        e for e in log
        if (since is None or e["timestamp"] >= since)
        and (until is None or e["timestamp"] < until)
    ]
    total = len(subset)
    referred = sum(1 for e in subset if e["status"] == "low_confidence")
    return {
        "total": total,
        "referred": referred,
        "rate": round(referred / total, 4) if total else None,
    }


def activity_by_month(log, years_back=4):
    """Son `years_back` yil icin yil x ay (12 hucre) etkinlik izgarasi.

    Her hucre o ay gelen toplam soruyu ve bunlarin kacinin insana
    yonlendirildigini tasir; sorumlu boylece hangi donemde ne kadar yogunluk
    oldugunu ve yogunlugun ne kadarinin cevapsiz kaldigini tek bakista gorur.
    Zaman damgasi 'YYYY-MM-DD HH:MM:SS' sabit genislikte oldugu icin yil/ay
    dilimlemeyle okunur - ayristirma maliyeti yok."""
    current_year = int(time.strftime("%Y"))
    first_year = current_year - years_back + 1

    grid = {
        year: [{"total": 0, "referred": 0} for _ in range(12)]
        for year in range(first_year, current_year + 1)
    }

    for entry in log:
        stamp = entry.get("timestamp") or ""
        if len(stamp) < 7:
            continue
        try:
            year = int(stamp[:4])
            month = int(stamp[5:7])
        except ValueError:
            continue
        if year not in grid or not 1 <= month <= 12:
            continue
        cell = grid[year][month - 1]
        cell["total"] += 1
        if entry.get("status") == "low_confidence":
            cell["referred"] += 1

    return [{"year": year, "months": grid[year]} for year in sorted(grid)]


def _cosine_clusters(vectors, threshold):
    """Kucuk N icin yeterince hizli, basit greedy kumeleme: ilk atanmamis
    ogenin vektoru kumenin 'temsilcisi' olur, esik ustundeki her atanmamis
    oge o kumeye eklenir."""
    n = len(vectors)
    assigned = [False] * n
    clusters = []
    for i in range(n):
        if assigned[i]:
            continue
        members = [i]
        assigned[i] = True
        for j in range(i + 1, n):
            if not assigned[j] and float(vectors[i] @ vectors[j]) >= threshold:
                members.append(j)
                assigned[j] = True
        clusters.append(members)
    return clusters


def frequent_unanswered(entries, min_count=FREQUENT_MIN_COUNT,
                        threshold=FREQUENT_SIMILARITY, top_n=FREQUENT_TOP_N):
    """entries: henuz SSS ile cevaplanmamis, kanit yetersizligiyle sorumluya
    yonlenen ('low_confidence') qa_log kayitlari - AI'in basariyla cevapladigi
    sorular DAHIL EDILMEZ. Her yarisma/baglam kendi icinde kumelenir, cunku
    ayni ifadeyle sorulan bir soru farkli yarismalar icin farkli cevap
    gerektirebilir."""
    by_scope = {}
    for e in entries:
        by_scope.setdefault(e.get("competition") or "Genel", []).append(e)

    results = []
    for scope, items in by_scope.items():
        if len(items) < min_count:
            continue
        vectors = embed_passages([it["question"] for it in items], batch_size=32, show_progress_bar=False)
        for members in _cosine_clusters(vectors, threshold):
            if len(members) < min_count:
                continue
            group = sorted((items[i] for i in members), key=lambda e: e["timestamp"], reverse=True)
            representative = group[0]["question"]
            variants = [g["question"] for g in group[1:] if g["question"] != representative]
            results.append({
                "question": representative,
                "competition": scope,
                "count": len(group),
                "last_asked": group[0]["timestamp"],
                "variants": variants[:6],
            })

    results.sort(key=lambda c: (c["count"], c["last_asked"]), reverse=True)
    return results[:top_n]

File: worker/src/lib/insights.ts
// backend/insights.py'nin D1 karsiligi. qa_log tablosu artik dosya yerine
// D1'de oldugu icin bazi hesaplamalar (referral_rate, activity_by_month)
// SQL ile yapilabilir; frequent_unanswered (embedding tabanli kumeleme)
// orijinaldeki gibi bellekte, ama RAG'de zaten kullanilan embedding
// modeliyle (embedQuery) yapiliyor.

import { embedQuery } from "./ai/embeddingProvider";
import { CONFIDENCE_HIGH } from "../config/rag";
import type { QaLogEntry, FeedbackEntry } from "./qaLog";

export function qualityBreakdown(log: QaLogEntry[]): { high: number; mid: number; total: number } {
  const answered = log.filter((e) => e.status === "answered" && e.top_score !== null);
  const high = answered.filter((e) => (e.top_score as number) > CONFIDENCE_HIGH).length;
  return { high, mid: answered.length - high, total: answered.length };
}

// backend/insights.py satisfaction_breakdown()'un birebir karsiligi:
// yarismacinin yanitin altindaki begen/begenme (thumbs up/down) ile
// DOGRUDAN bildirdigi memnuniyet - quality_breakdown'daki model-guven
// vekilinden farkli olarak kullanicinin kendi degerlendirmesidir.
export function satisfactionBreakdown(
  feedback: FeedbackEntry[],
): { up: number; down: number; total: number; rate: number | null } {
  const up = feedback.filter((f) => f.satisfaction === "up").length;
  const down = feedback.filter((f) => f.satisfaction === "down").length;
  const total = up + down;
  return { up, down, total, rate: total ? Math.round((up / total) * 10000) / 10000 : null };
}

export function recentCutoff(days: number): string {
  const d = new Date(Date.now() - days * 86400 * 1000);
  return d.toISOString().slice(0, 19).replace("T", " ");
}

export function referralRate(
  log: QaLogEntry[],
  since?: string,
  until?: string,
): { total: number; referred: number; rate: number | null } {
  const subset = log.filter(
    (e) => (since === undefined || e.timestamp >= since) && (until === undefined || e.timestamp < until),
  );
  const total = subset.length;
  const referred = subset.filter((e) => e.status === "low_confidence").length;
  return { total, referred, rate: total ? Math.round((referred / total) * 10000) / 10000 : null };
}

export function activityByMonth(
  log: QaLogEntry[],
  yearsBack = 4,
): Array<{ year: number; months: Array<{ total: number; referred: number }> }> {
  const currentYear = new Date().getFullYear();
  const firstYear = currentYear - yearsBack + 1;

  const grid = new Map<number, Array<{ total: number; referred: number }>>();
  for (let y = firstYear; y <= currentYear; y++) {
    grid.set(
      y,
      Array.from({ length: 12 }, () => ({ total: 0, referred: 0 })),
    );
  }

  for (const entry of log) {
    const stamp = entry.timestamp || "";
    if (stamp.length < 7) continue;
    const year = parseInt(stamp.slice(0, 4), 10);
    const month = parseInt(stamp.slice(5, 7), 10);
    if (!grid.has(year) || month < 1 || month > 12) continue;
    const cell = grid.get(year)![month - 1];
    cell.total += 1;
    if (entry.status === "low_confidence") cell.referred += 1;
  }

  return [...grid.keys()].sort().map((year) => ({ year, months: grid.get(year)! }));
}

const FREQUENT_MIN_COUNT = 2;
const FREQUENT_SIMILARITY = 0.86;
const FREQUENT_TOP_N = 20;

function dot(a: number[], b: number[]): number {
  let s = 0;
  for (let i = 0; i < a.length; i++) s += a[i] * b[i];
  return s;
}

function cosineClusters(vectors: number[][], threshold: number): number[][] {
  const n = vectors.length;
  const assigned = new Array(n).fill(false);
  const clusters: number[][] = [];
  for (let i = 0; i < n; i++) {
    if (assigned[i]) continue;
    const members = [i];
    assigned[i] = true;
    for (let j = i + 1; j < n; j++) {
      if (!assigned[j] && dot(vectors[i], vectors[j]) >= threshold) {
        members.push(j);
        assigned[j] = true;
      }
    }
    clusters.push(members);
  }
  return clusters;
}

export interface FrequentGroup {
  question: string;
  competition: string;
  count: number;
  last_asked: string;
  variants: string[];
}

// backend/insights.py _frequent_clusters()'in birebir karsiligi — hem
// frequent_unanswered (sadece insana yonlenenler) hem frequent_topics
// (basariyla cevaplananlar dahil TUM trafik) tarafindan paylasilir. Her
// yarisma/baglam kendi icinde kumelenir, cunku ayni ifadeyle sorulan bir
// soru farkli yarismalar icin farkli cevap gerektirebilir.
async function frequentClusters(
  ai: Ai,
  entries: QaLogEntry[],
  minCount: number,
  threshold: number,
  topN: number,
): Promise<FrequentGroup[]> {
  const byScope = new Map<string, QaLogEntry[]>();
  for (const e of entries) {
    const key = e.competition || "Genel";
    if (!byScope.has(key)) byScope.set(key, []);
    byScope.get(key)!.push(e);
  }

  const results: FrequentGroup[] = [];
  for (const [scope, items] of byScope) {
    if (items.length < minCount) continue;
    // embed_passages'in aksine embedQuery tek metin alir — burada da RAG'in
    // ayni embedding modeli kullanildigi icin (paralel) tek tek cagriliyor.
    const vectors = await Promise.all(items.map((it) => embedQuery(ai, it.question)));
    for (const members of cosineClusters(vectors, threshold)) {
      if (members.length < minCount) continue;
      const group = members
        .map((i) => items[i])
        .sort((a, b) => (a.timestamp < b.timestamp ? 1 : -1));
      const representative = group[0].question;
      const variants = group
        .slice(1)
        .map((g) => g.question)
        .filter((q) => q !== representative);
      results.push({
        question: representative,
        competition: scope,
        count: group.length,
        last_asked: group[0].timestamp,
        variants: [...new Set(variants)].slice(0, 6),
      });
    }
  }

  results.sort((a, b) => (b.count !== a.count ? b.count - a.count : (b.last_asked > a.last_asked ? 1 : -1)));
  return results.slice(0, topN);
}

// frequent_unanswered()'in birebir karsiligi. entries: henuz SSS ile
// cevaplanmamis, 'low_confidence' qa_log kayitlari - AI'in basariyla
// cevapladigi sorular DAHIL EDILMEZ.
export function frequentUnanswered(
  ai: Ai,
  entries: QaLogEntry[],
  minCount = FREQUENT_MIN_COUNT,
  threshold = FREQUENT_SIMILARITY,
  topN = FREQUENT_TOP_N,
): Promise<FrequentGroup[]> {
  return frequentClusters(ai, entries, minCount, threshold, topN);
}

// backend/insights.py frequent_topics()'in birebir karsiligi:
// frequent_unanswered'in aksine SADECE insana yonlenen degil, basariyla
// cevaplanan sorular DAHIL TUM soru trafigi uzerinden kumeler - sorumlunun
// yarismacilarin GENEL OLARAK en cok hangi konularda soru sordugunu
// gormesi icindir. 'unclear' (anlamsiz/klavye karalamasi) kayitlar gercek
// bir soru olmadigindan disarida birakilir.
export function frequentTopics(
  ai: Ai,
  log: QaLogEntry[],
  minCount = FREQUENT_MIN_COUNT,
  threshold = FREQUENT_SIMILARITY,
  topN = FREQUENT_TOP_N,
): Promise<FrequentGroup[]> {
  const entries = log.filter((e) => e.status !== "unclear");
  return frequentClusters(ai, entries, minCount, threshold, topN);
}
